### PR TITLE
azuredevops: Add client method ListAuthorizedUserOrganizations

### PR DESCRIPTION
**Pull request by indradhanush**
First part of #48555.

## Test plan

- Added tests
- Builds should pass

### CODE REVIEW

File: internal/extsvc/azuredevops/users.go [hunk 1]
```diff
@@ -41,6 +41,31 @@ func (c *client) GetAuthorizedProfile(ctx context.Context) (Profile, error) {
...
+
+	apiURL := VisualStudioAppURL
+	if MockVisualStudioAppURL != "" {
+		apiURL = MockVisualStudioAppURL
+	}
+
+	response := ListAuthorizedUserOrgsResponse{}
+	if _, err := c.do(ctx, req, apiURL, &response); err != nil {
+		return nil, err
+	}
```
Comment (indradhanush) [inline internal/extsvc/azuredevops/users.go:64]: We probably want to support pagination here. But I'm not sure I know of a good way to test that code. As a result I've left it out of the initial version and can follow up with an improvement if required.

Comment (varsanojidan) [reply internal/extsvc/azuredevops/users.go:64]: GetRepositoryBranch in repositories.go has an example of pagination with the ADO API. You can limit you results with the $top query param to see that the pagination works.

File: internal/extsvc/azuredevops/users_test.go [hunk 2]
```diff
@@ -18,3 +18,37 @@ func TestClient_AzureServicesProfile(t *testing.T) {
 
 	testutil.AssertGolden(t, "testdata/golden/AzureServicesProfile.json", *update, resp)
 }
+
+// To update this test run:
+//  1. Set the env AZURE_DEV_OPS_USERNAME and AZURE_DEV_OPS_TOKEN (the secrets can be found in 1Password if you search for Azure test credentials)
+//  2. Run the test with the -update flag:
+//     `go test -run='TestClient_ListAuthorizedUserOrganizations' -update=true`
+func TestClient_ListAuthorizedUserOrganizations(t *testing.T) {
+	// cf, save := newClientFactory(t, t.Name())
```
Comment (varsanojidan) [inline internal/extsvc/azuredevops/users_test.go:27]: I think this might be leftover?

File: internal/extsvc/azuredevops/users.go [hunk 3]
```diff
@@ -41,6 +41,31 @@ func (c *client) GetAuthorizedProfile(ctx context.Context) (Profile, error) {
 	return p, nil
 }
 
+func (c *client) ListAuthorizedUserOrganizations(ctx context.Context, profile Profile) ([]Org, error) {
+	reqURL := url.URL{Path: "_apis/accounts"}
```
Comment (varsanojidan) [inline internal/extsvc/azuredevops/users.go:45]: Can you add a call to IsAzureDevOpsServices() here and return an error if it isn't?

**Review by varsanojidan — APPROVED** [review]
Looks good 😄, left some minor comments.

File: internal/extsvc/azuredevops/users.go [hunk 4]
```diff
@@ -41,6 +41,31 @@ func (c *client) GetAuthorizedProfile(ctx context.Context) (Profile, error) {
...
+	if err != nil {
+		return nil, err
+	}
+
+	queryParams := req.URL.Query()
+	queryParams.Set("memberId", profile.PublicAlias)
+	req.URL.RawQuery = queryParams.Encode()
+
+	apiURL := VisualStudioAppURL
+	if MockVisualStudioAppURL != "" {
```
Comment (varsanojidan) [inline internal/extsvc/azuredevops/users.go:57]: is this needed here?

File: internal/extsvc/azuredevops/users_test.go [hunk 5]
```diff
@@ -18,3 +18,37 @@ func TestClient_AzureServicesProfile(t *testing.T) {
 
 	testutil.AssertGolden(t, "testdata/golden/AzureServicesProfile.json", *update, resp)
 }
+
+// To update this test run:
+//  1. Set the env AZURE_DEV_OPS_USERNAME and AZURE_DEV_OPS_TOKEN (the secrets can be found in 1Password if you search for Azure test credentials)
```
Comment (cbart) [inline internal/extsvc/azuredevops/users_test.go:23]: How does this run through CI?

**Review by cbart — APPROVED** [review]

File: internal/extsvc/azuredevops/users_test.go [hunk 6]
```diff
@@ -18,3 +18,37 @@ func TestClient_AzureServicesProfile(t *testing.T) {
...
+	// if err != nil {
+	// 	t.Fatalf("failed to created http client: %v", err)
+	// }
+
+	cli, save := NewTestClient(
+		t,
+		"ListAuthorizedUserOrganizations",
+		*update,
+	)
+	defer save()
```
Comment (sashaostrikov) [inline internal/extsvc/azuredevops/users_test.go:40]: nit:
```suggestion
	t.Cleanup(save)
```

File: internal/extsvc/azuredevops/users_test.go [hunk 7]
```diff
@@ -18,3 +18,37 @@ func TestClient_AzureServicesProfile(t *testing.T) {
...
+
+	ctx := context.Background()
+	profile, err := cli.GetAuthorizedProfile(ctx)
+	if err != nil {
+		t.Fatalf("failed to authorized profile: %v", err)
```
Comment (sashaostrikov) [inline internal/extsvc/azuredevops/users_test.go:45]:
```suggestion
		t.Fatalf("failed to get authorized profile: %v", err)
```

**Review by sashaostrikov — APPROVED** [review]
LGTM!

File: internal/extsvc/azuredevops/users.go (hunk 4 above)
Comment (indradhanush) [reply internal/extsvc/azuredevops/users.go:57]: Not yet, but I will need it in tests as soon as I use this in permissions sync (#48574).

File: internal/extsvc/azuredevops/users.go (hunk 1 above)
Comment (indradhanush) [reply internal/extsvc/azuredevops/users.go:69]: Hmm. I checked the [docs](https://learn.microsoft.com/en-us/rest/api/azure/devops/account/accounts/list?view=azure-devops-rest-7.0&tabs=HTTP#uri-parameters) and it does not seem like this API endpoint supports `continuationToken` and `top` like some others, for example: https://learn.microsoft.com/en-us/rest/api/azure/devops/git/refs/list?view=azure-devops-rest-6.0&tabs=HTTP#uri-parameters.

Also, in my testing adding `$top=1` had no effect on this API either.

File: internal/extsvc/azuredevops/users_test.go (hunk 5 above)
Comment (indradhanush) [reply internal/extsvc/azuredevops/users_test.go:23]: By default, the test will replay the request from the [vcr](https://github.com/sourcegraph/sourcegraph/pull/48572/files#diff-7e795d68ea437050dd9fcdcfb7253b34bb88c3af0fda664ce84323e41457c009) file and use the response stored in it to generate a mock HTTP response. The test assertion will then use the contents of the [golden](https://github.com/sourcegraph/sourcegraph/pull/48572/files#diff-fcca986b92fcfc66079105dcd12922fa5733f4f0d5b12248cb567a9841b71e2f) for the assertion.

But if the `-update=true` flag is set then this test will skip the vcr and send an actual HTTP request and update the contents of the vcr file. 

So if you ran the test with `-update=true` without setting the token as documented, you'd see a bunch of 401s in the vcr file.

**Pull request merged**